File: sdk/async-scala/src/main/scala/io/swagger/client/api/SpecApi.scala
package io.swagger.client.api

import io.swagger.client.model.Any
import io.swagger.client.model.Spec
import io.swagger.client.model.SpecOption
import io.swagger.client.model.SpecProductAssignment
import com.wordnik.swagger.client._
import scala.concurrent.Future
import collection.mutable

class SpecApi(client: TransportClient, config: SwaggerConfig) extends ApiClient(client, config) {

  def create(spec: Spec)(implicit reader: ClientResponseReader[Any], writer: RequestWriter[Spec]): Future[Any] = {
    // create path and map variables
    val path = (addFmt("/specs"))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]

    if (spec == null) throw new Exception("Missing required parameter 'spec' when calling SpecApi->create")



    val resFuture = client.submit("POST", path, queryParams.toMap, headerParams.toMap, writer.write(spec))
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def createOption(specID: String,
      option: SpecOption)(implicit reader: ClientResponseReader[Any], writer: RequestWriter[SpecOption]): Future[Any] = {
    // create path and map variables
    val path = (addFmt("/specs/{specID}/options")
        replaceAll ("\\{" + "specID" + "\\}",specID.toString))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]

    if (option == null) throw new Exception("Missing required parameter 'option' when calling SpecApi->createOption")



    val resFuture = client.submit("POST", path, queryParams.toMap, headerParams.toMap, writer.write(option))
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def delete(specID: String)(implicit reader: ClientResponseReader[Unit]): Future[Unit] = {
    // create path and map variables
    val path = (addFmt("/specs/{specID}")
        replaceAll ("\\{" + "specID" + "\\}",specID.toString))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]



    val resFuture = client.submit("DELETE", path, queryParams.toMap, headerParams.toMap, "")
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def deleteOption(specID: String,
      optionID: String)(implicit reader: ClientResponseReader[Unit]): Future[Unit] = {
    // create path and map variables
    val path = (addFmt("/specs/{specID}/options/{optionID}")
        replaceAll ("\\{" + "specID" + "\\}",specID.toString)
        replaceAll ("\\{" + "optionID" + "\\}",optionID.toString))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]



    val resFuture = client.submit("DELETE", path, queryParams.toMap, headerParams.toMap, "")
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def deleteProductAssignment(specID: String,
      productID: String)(implicit reader: ClientResponseReader[Unit]): Future[Unit] = {
    // create path and map variables
    val path = (addFmt("/specs/{specID}/productassignments/{productID}")
        replaceAll ("\\{" + "specID" + "\\}",specID.toString)
        replaceAll ("\\{" + "productID" + "\\}",productID.toString))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]



    val resFuture = client.submit("DELETE", path, queryParams.toMap, headerParams.toMap, "")
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def get(specID: String)(implicit reader: ClientResponseReader[Any]): Future[Any] = {
    // create path and map variables
    val path = (addFmt("/specs/{specID}")
        replaceAll ("\\{" + "specID" + "\\}",specID.toString))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]



    val resFuture = client.submit("GET", path, queryParams.toMap, headerParams.toMap, "")
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def getOption(specID: String,
      optionID: String)(implicit reader: ClientResponseReader[Any]): Future[Any] = {
    // create path and map variables
    val path = (addFmt("/specs/{specID}/options/{optionID}")
        replaceAll ("\\{" + "specID" + "\\}",specID.toString)
        replaceAll ("\\{" + "optionID" + "\\}",optionID.toString))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]



    val resFuture = client.submit("GET", path, queryParams.toMap, headerParams.toMap, "")
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def list(search: Option[String] = None,
      searchOn: Option[String] = None,
      sortBy: Option[String] = None,
      page: Option[Integer] = None,
      pageSize: Option[Integer] = None
      )(implicit reader: ClientResponseReader[Any]): Future[Any] = {
    // create path and map variables
    val path = (addFmt("/specs"))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]

    if (search != null) search.foreach { v => queryParams += "search" -> v.toString }

    if (searchOn != null) searchOn.foreach { v => queryParams += "searchOn" -> v.toString }

    if (sortBy != null) sortBy.foreach { v => queryParams += "sortBy" -> v.toString }

    if (page != null) page.foreach { v => queryParams += "page" -> v.toString }

    if (pageSize != null) pageSize.foreach { v => queryParams += "pageSize" -> v.toString }



    val resFuture = client.submit("GET", path, queryParams.toMap, headerParams.toMap, "")
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def listOptions(specID: String,
      search: Option[String] = None,
      searchOn: Option[String] = None,
      sortBy: Option[String] = None,
      page: Option[Integer] = None,
      pageSize: Option[Integer] = None
      )(implicit reader: ClientResponseReader[Any]): Future[Any] = {
    // create path and map variables
    val path = (addFmt("/specs/{specID}/options")
        replaceAll ("\\{" + "specID" + "\\}",specID.toString))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]

    if (search != null) search.foreach { v => queryParams += "search" -> v.toString }

    if (searchOn != null) searchOn.foreach { v => queryParams += "searchOn" -> v.toString }

    if (sortBy != null) sortBy.foreach { v => queryParams += "sortBy" -> v.toString }

    if (page != null) page.foreach { v => queryParams += "page" -> v.toString }

    if (pageSize != null) pageSize.foreach { v => queryParams += "pageSize" -> v.toString }



    val resFuture = client.submit("GET", path, queryParams.toMap, headerParams.toMap, "")
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def listProductAssignments(specID: Option[String] = None,
      productID: Option[String] = None,
      page: Option[Integer] = None,
      pageSize: Option[Integer] = None
      )(implicit reader: ClientResponseReader[Any]): Future[Any] = {
    // create path and map variables
    val path = (addFmt("/specs/productassignments"))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]

    if (specID != null) specID.foreach { v => queryParams += "specID" -> v.toString }

    if (productID != null) productID.foreach { v => queryParams += "productID" -> v.toString }

    if (page != null) page.foreach { v => queryParams += "page" -> v.toString }

    if (pageSize != null) pageSize.foreach { v => queryParams += "pageSize" -> v.toString }



    val resFuture = client.submit("GET", path, queryParams.toMap, headerParams.toMap, "")
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def patch(specID: String,
      spec: Spec)(implicit reader: ClientResponseReader[Any], writer: RequestWriter[Spec]): Future[Any] = {
    // create path and map variables
    val path = (addFmt("/specs/{specID}")
        replaceAll ("\\{" + "specID" + "\\}",specID.toString))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]

    if (spec == null) throw new Exception("Missing required parameter 'spec' when calling SpecApi->patch")



    val resFuture = client.submit("PATCH", path, queryParams.toMap, headerParams.toMap, writer.write(spec))
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def patchOption(specID: String,
      optionID: String,
      option: SpecOption)(implicit reader: ClientResponseReader[Any], writer: RequestWriter[SpecOption]): Future[Any] = {
    // create path and map variables
    val path = (addFmt("/specs/{specID}/options/{optionID}")
        replaceAll ("\\{" + "specID" + "\\}",specID.toString)
        replaceAll ("\\{" + "optionID" + "\\}",optionID.toString))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]

    if (option == null) throw new Exception("Missing required parameter 'option' when calling SpecApi->patchOption")



    val resFuture = client.submit("PATCH", path, queryParams.toMap, headerParams.toMap, writer.write(option))
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def saveProductAssignment(productAssignment: SpecProductAssignment)(implicit reader: ClientResponseReader[Unit], writer: RequestWriter[SpecProductAssignment]): Future[Unit] = {
    // create path and map variables
    val path = (addFmt("/specs/productassignments"))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]

    if (productAssignment == null) throw new Exception("Missing required parameter 'productAssignment' when calling SpecApi->saveProductAssignment")



    val resFuture = client.submit("POST", path, queryParams.toMap, headerParams.toMap, writer.write(productAssignment))
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def update(specID: String,
      spec: Spec)(implicit reader: ClientResponseReader[Any], writer: RequestWriter[Spec]): Future[Any] = {
    // create path and map variables
    val path = (addFmt("/specs/{specID}")
        replaceAll ("\\{" + "specID" + "\\}",specID.toString))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]

    if (spec == null) throw new Exception("Missing required parameter 'spec' when calling SpecApi->update")



    val resFuture = client.submit("PUT", path, queryParams.toMap, headerParams.toMap, writer.write(spec))
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }

  def updateOption(specID: String,
      optionID: String,
      option: SpecOption)(implicit reader: ClientResponseReader[Any], writer: RequestWriter[SpecOption]): Future[Any] = {
    // create path and map variables
    val path = (addFmt("/specs/{specID}/options/{optionID}")
        replaceAll ("\\{" + "specID" + "\\}",specID.toString)
        replaceAll ("\\{" + "optionID" + "\\}",optionID.toString))

    // query params
    val queryParams = new mutable.HashMap[String, String]
    val headerParams = new mutable.HashMap[String, String]

    if (option == null) throw new Exception("Missing required parameter 'option' when calling SpecApi->updateOption")



    val resFuture = client.submit("PUT", path, queryParams.toMap, headerParams.toMap, writer.write(option))
    resFuture flatMap { resp =>
      process(reader.read(resp))
    }
  }


}

File: sdk/sinatra/api/spec_api.rb
require 'json'


MyApp.add_route('POST', '/v1/specs', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "create", 
  "responseClass" => "object", 
  "endpoint" => "/specs", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "body",
      "description" => "",
      "dataType" => "Spec",
      "paramType" => "body",
    }
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('POST', '/v1/specs/{specID}/options', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "create_option", 
  "responseClass" => "object", 
  "endpoint" => "/specs/{specID}/options", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "spec_id",
      "description" => "ID of the spec.",
      "dataType" => "string",
      "paramType" => "path",
    },
    {
      "name" => "body",
      "description" => "",
      "dataType" => "SpecOption",
      "paramType" => "body",
    }
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('DELETE', '/v1/specs/{specID}', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "delete", 
  "responseClass" => "void", 
  "endpoint" => "/specs/{specID}", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "spec_id",
      "description" => "ID of the spec.",
      "dataType" => "string",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('DELETE', '/v1/specs/{specID}/options/{optionID}', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "delete_option", 
  "responseClass" => "void", 
  "endpoint" => "/specs/{specID}/options/{optionID}", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "spec_id",
      "description" => "ID of the spec.",
      "dataType" => "string",
      "paramType" => "path",
    },
    {
      "name" => "option_id",
      "description" => "ID of the option.",
      "dataType" => "string",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('DELETE', '/v1/specs/{specID}/productassignments/{productID}', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "delete_product_assignment", 
  "responseClass" => "void", 
  "endpoint" => "/specs/{specID}/productassignments/{productID}", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "spec_id",
      "description" => "ID of the spec.",
      "dataType" => "string",
      "paramType" => "path",
    },
    {
      "name" => "product_id",
      "description" => "ID of the product.",
      "dataType" => "string",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('GET', '/v1/specs/{specID}', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "get", 
  "responseClass" => "object", 
  "endpoint" => "/specs/{specID}", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "spec_id",
      "description" => "ID of the spec.",
      "dataType" => "string",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('GET', '/v1/specs/{specID}/options/{optionID}', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "get_option", 
  "responseClass" => "object", 
  "endpoint" => "/specs/{specID}/options/{optionID}", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "spec_id",
      "description" => "ID of the spec.",
      "dataType" => "string",
      "paramType" => "path",
    },
    {
      "name" => "option_id",
      "description" => "ID of the option.",
      "dataType" => "string",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('GET', '/v1/specs', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "list", 
  "responseClass" => "object", 
  "endpoint" => "/specs", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "search",
      "description" => "Word or phrase to search for.",
      "dataType" => "string",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    {
      "name" => "search_on",
      "description" => "Comma-delimited list of fields to search on.",
      "dataType" => "string",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    {
      "name" => "sort_by",
      "description" => "Comma-delimited list of fields to sort by.",
      "dataType" => "string",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    {
      "name" => "page",
      "description" => "Page of results to return. Default: 1",
      "dataType" => "int",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    {
      "name" => "page_size",
      "description" => "Number of results to return per page. Default: 20, max: 100.",
      "dataType" => "int",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('GET', '/v1/specs/{specID}/options', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "list_options", 
  "responseClass" => "object", 
  "endpoint" => "/specs/{specID}/options", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "search",
      "description" => "Word or phrase to search for.",
      "dataType" => "string",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    {
      "name" => "search_on",
      "description" => "Comma-delimited list of fields to search on.",
      "dataType" => "string",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    {
      "name" => "sort_by",
      "description" => "Comma-delimited list of fields to sort by.",
      "dataType" => "string",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    {
      "name" => "page",
      "description" => "Page of results to return. Default: 1",
      "dataType" => "int",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    {
      "name" => "page_size",
      "description" => "Number of results to return per page. Default: 20, max: 100.",
      "dataType" => "int",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    {
      "name" => "spec_id",
      "description" => "ID of the spec.",
      "dataType" => "string",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('GET', '/v1/specs/productassignments', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "list_product_assignments", 
  "responseClass" => "object", 
  "endpoint" => "/specs/productassignments", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "spec_id",
      "description" => "ID of the spec.",
      "dataType" => "string",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    {
      "name" => "product_id",
      "description" => "ID of the product.",
      "dataType" => "string",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    {
      "name" => "page",
      "description" => "Page of results to return. Default: 1",
      "dataType" => "int",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    {
      "name" => "page_size",
      "description" => "Number of results to return per page. Default: 20, max: 100.",
      "dataType" => "int",
      "paramType" => "query",
      
      "allowableValues" => "",
      
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('PATCH', '/v1/specs/{specID}', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "patch", 
  "responseClass" => "object", 
  "endpoint" => "/specs/{specID}", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "spec_id",
      "description" => "ID of the spec.",
      "dataType" => "string",
      "paramType" => "path",
    },
    {
      "name" => "body",
      "description" => "",
      "dataType" => "Spec",
      "paramType" => "body",
    }
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('PATCH', '/v1/specs/{specID}/options/{optionID}', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "patch_option", 
  "responseClass" => "object", 
  "endpoint" => "/specs/{specID}/options/{optionID}", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "spec_id",
      "description" => "ID of the spec.",
      "dataType" => "string",
      "paramType" => "path",
    },
    {
      "name" => "option_id",
      "description" => "ID of the option.",
      "dataType" => "string",
      "paramType" => "path",
    },
    {
      "name" => "body",
      "description" => "",
      "dataType" => "SpecOption",
      "paramType" => "body",
    }
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('POST', '/v1/specs/productassignments', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "save_product_assignment", 
  "responseClass" => "void", 
  "endpoint" => "/specs/productassignments", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "body",
      "description" => "",
      "dataType" => "SpecProductAssignment",
      "paramType" => "body",
    }
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('PUT', '/v1/specs/{specID}', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "update", 
  "responseClass" => "object", 
  "endpoint" => "/specs/{specID}", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "spec_id",
      "description" => "ID of the spec.",
      "dataType" => "string",
      "paramType" => "path",
    },
    {
      "name" => "body",
      "description" => "",
      "dataType" => "Spec",
      "paramType" => "body",
    }
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('PUT', '/v1/specs/{specID}/options/{optionID}', {
  "resourcePath" => "/Spec",
  "summary" => "",
  "nickname" => "update_option", 
  "responseClass" => "object", 
  "endpoint" => "/specs/{specID}/options/{optionID}", 
  "notes" => "",
  "parameters" => [
    {
      "name" => "spec_id",
      "description" => "ID of the spec.",
      "dataType" => "string",
      "paramType" => "path",
    },
    {
      "name" => "option_id",
      "description" => "ID of the option.",
      "dataType" => "string",
      "paramType" => "path",
    },
    {
      "name" => "body",
      "description" => "",
      "dataType" => "SpecOption",
      "paramType" => "body",
    }
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end

